### PR TITLE
feat: add slack trigger

### PR DESCRIPTION
**Pull request by chaseterry**
Created a Slack trigger and added a tutorial. Closing #545

### CODE REVIEW

File: controllers/sensor/validate.go [hunk 1]
```diff
@@ -321,6 +326,20 @@ func validateNATSTrigger(trigger *v1alpha1.NATSTrigger) error {
...
 
+// validateSlackTrigger validates the Slack trigger.
+func validateSlackTrigger(trigger *v1alpha1.SlackTrigger) error {
+	if trigger == nil {
+		return errors.New("trigger can't be nil")
+	}
+	if trigger.SlackToken == nil {
+		return errors.New("slack token can't be empty")
+	}
+	if trigger.Namespace == "" {
```
Comment (VaibhavPage) [inline controllers/sensor/validate.go:337]: is this required? The namespace should fall back to gateway's namespace if not specifically defined by user.

File: docs/triggers/slack-trigger.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+**_Note_**: If you define both the `contextKey` and `dataKey` within a paramter item, then
+the `dataKey` takes the precedence.
+
+You can create any paramater structure you want. To get more info on how to 
+generate complex event payloads, take a look at [this library](https://github.com/tidwall/sjson).
+
+The complete specification of Slack trigger is available [here](https://github.com/argoproj/argo-events/blob/master/api/sensor.md#slacktrigger).
+
+### Prerequisite
```
Comment (VaibhavPage) [inline docs/triggers/slack-trigger.md:106]: Remove the Kubeless prerequisite

File: pkg/apis/sensor/v1alpha1/types.go [hunk 3]
```diff
@@ -520,6 +523,25 @@ type CustomTrigger struct {
...
+	// Parameters is the list of key-value extracted from event's payload that are applied to
+	// the trigger resource.
+	// +listType=triggerParameters
+	// +optional
+	Parameters []TriggerParameter `json:"parameters,omitempty" protobuf:"bytes,1,rep,name=parameters"`
+	// SlackToken refers to the Kubernetes secret that holds the slack token required to send messages.
+	SlackToken *corev1.SecretKeySelector `json:"slackToken" protobuf:"bytes,2,name=slackToken"`
+	// Namespace to read the password secret from.
+	// This is required if the password secret selector is specified.
+	Namespace string `json:"namespace" protobuf:"bytes,3,name=namespace"`
```
Comment (VaibhavPage) [inline pkg/apis/sensor/v1alpha1/types.go:536]: make `Namespace` optional.

File: sensors/triggers/slack/slack.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package slack
+
+import (
+	//"bytes"
```
Comment (VaibhavPage) [inline sensors/triggers/slack/slack.go:19]: remove commented imports and run `go import` to organize imports

File: sensors/triggers/slack/slack.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+		return nil, errors.New("failed to marshal the Slack trigger resource")
+	}
+
+	slacktrigger := t.Trigger.Template.Slack
+
+	channel := slacktrigger.Channel
+	if channel == "" {
+		return nil, errors.New("no slack channel provided")
+	}
+
```
Comment (VaibhavPage) [inline sensors/triggers/slack/slack.go:106]: insert an if condition to set `namespace` for `slacktrigger` if not explicitly defined by the user. Take a look at other gateways that deals with K8s secrets and namespace.

File: sensors/triggers/slack/slack.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	message := slacktrigger.Message
+	if message == "" {
+		return nil, errors.New("no slack message to post")
+	}
+
+	slackToken := ""
+	if slacktrigger.SlackToken != nil {
+		secretToken, err := common.GetSecrets(t.K8sClient, slacktrigger.Namespace, slacktrigger.SlackToken)
```
Comment (VaibhavPage) [inline sensors/triggers/slack/slack.go:114]: remove `secretToken` variable, just assign `slackToken` the value of secret data,

`slackToken, err := common.GetSecrets(t.K8sClient, slacktrigger.Namespace, slacktrigger.SlackToken)`

File: sensors/triggers/slack/slack_test.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	assert.Nil(t, err)
+	assert.NotNil(t, resource)
+
+	ot, ok := resource.(*v1alpha1.SlackTrigger)
+	assert.Equal(t, true, ok)
+	assert.Equal(t, "real-channel", ot.Channel)
+	assert.Equal(t, "real-message", ot.Message)
+}
+
+//func TestSlackTrigger_Execute(t *testing.T) {
```
Comment (VaibhavPage) [inline sensors/triggers/slack/slack_test.go:131]: remove commented code.

File: docs/triggers/slack-trigger.md (hunk 2 above)
Comment (chaseterry) [reply docs/triggers/slack-trigger.md]: Removed

File: controllers/sensor/validate.go (hunk 1 above)
Comment (chaseterry) [reply controllers/sensor/validate.go]: Remove required namespace, and made trigger's namespace fall back to sensor's namespace

File: pkg/apis/sensor/v1alpha1/types.go (hunk 3 above)
Comment (chaseterry) [reply pkg/apis/sensor/v1alpha1/types.go]: Made Namespace optional

File: sensors/triggers/slack/slack.go (hunk 4 above)
Comment (chaseterry) [reply sensors/triggers/slack/slack.go]: cleaned up imports

File: sensors/triggers/slack/slack.go (hunk 5 above)
Comment (chaseterry) [reply sensors/triggers/slack/slack.go:102]: Made trigger's namespace fall back to sensor's namespace

File: sensors/triggers/slack/slack_test.go (hunk 7 above)
Comment (chaseterry) [reply sensors/triggers/slack/slack_test.go]: Removed

**Pull request merged**